qa/rpc-tests/wallet_listnotes.py

```diff
@@ -0,0 +1,168 @@
+        
+        # No funds in saplingzaddr yet
+        assert_equal(0, len(self.nodes[0].z_listunspent(0, 9999, False, [saplingzaddr])))
+
+        # Send 0.9999 to our sapling zaddr
+        # (sending from a sprout zaddr to a sapling zaddr is disallowed,
+        # so send from coin base)
+        receive_amount_2 = Decimal('2.0') - Decimal('0.0001')
+        recipients = [{"address": saplingzaddr, "amount":receive_amount_2}]
+        myopid = self.nodes[0].z_sendmany(mining_addr, recipients)
+        txid_3 = wait_and_assert_operationid_status(self.nodes[0], myopid)
+        self.sync_all()
+        unspent_tx = self.nodes[0].z_listunspent(0)
+        assert_equal(3, len(unspent_tx))
+
+        # low-to-high in amount
+        unspent_tx = sorted(unspent_tx, key=lambda k: k['amount'])
+
+        assert_equal(False,             unspent_tx[0]['change'])
+        assert_equal(txid_2,            unspent_tx[0]['txid'])
+        assert_equal(True,              unspent_tx[0]['spendable'])
+        assert_equal(sproutzaddr2,      unspent_tx[0]['address'])
+        assert_equal(receive_amount_1,  unspent_tx[0]['amount'])
+
+        assert_equal(False,             unspent_tx[1]['change'])
+        assert_equal(txid_3,            unspent_tx[1]['txid'])
+        assert_equal(True,              unspent_tx[1]['spendable'])
+        assert_equal(saplingzaddr,      unspent_tx[1]['address'])
+        assert_equal(receive_amount_2,  unspent_tx[1]['amount'])
+
+        assert_equal(True,              unspent_tx[2]['change'])
+        assert_equal(txid_2,            unspent_tx[2]['txid'])
+        assert_equal(True,              unspent_tx[2]['spendable'])
+        assert_equal(sproutzaddr,       unspent_tx[2]['address'])
+        assert_equal(change_amount_9,   unspent_tx[2]['amount'])
+
+        unspent_tx_filter = self.nodes[0].z_listunspent(0, 9999, False, [saplingzaddr])
+        assert_equal(1, len(unspent_tx_filter))
+        assert_equal(unspent_tx[1], unspent_tx_filter[0])
+
+        # test that pre- and post-sapling can be filtered in a single call
+        unspent_tx_filter = self.nodes[0].z_listunspent(0, 9999, False,
+            [sproutzaddr, saplingzaddr])
+        assert_equal(2, len(unspent_tx_filter))
+        unspent_tx_filter = sorted(unspent_tx_filter, key=lambda k: k['amount'])
+        assert_equal(unspent_tx[1], unspent_tx_filter[0])
+        assert_equal(unspent_tx[2], unspent_tx_filter[1])
+
+        # so far, this node has no watchonly addresses, so results are the same
+        unspent_tx_watchonly = self.nodes[0].z_listunspent(0, 9999, True)
+        unspent_tx_watchonly = sorted(unspent_tx_watchonly, key=lambda k: k['amount'])
+        assert_equal(unspent_tx, unspent_tx_watchonly)
+
+        # TODO: use z_exportviewingkey, z_importviewingkey to test includeWatchonly
+        # but this requires Sapling support for those RPCs
+
+if __name__ == '__main__':
+    WalletListNotes().main()
```

Comment:
My comment on this test is that we seem to be testing a lot of the sprout logic (Which I would expect to be tested elsewhere). We do not test the sapling note change logic (but I think this is tested as part of the update to z_listreceivedbyarrdess), we do not test that we use outindex rather than jsindex/jsoutindex, and we do not test the memo field. These are the differences between sprout/sapling, and therefore, I was expecting them to be tested.

That being said, in the cpp file these look good. Please look at the comment regarding the memo and once that is verified this should be ackable.

Reply:
Filed https://github.com/zcash/zcash/issues/3562